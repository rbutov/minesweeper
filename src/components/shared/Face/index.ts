enum Faces {
  UNPRESSED = 'unpressed',
  LOSE = 'lose',
  WIN = 'win',
}

type FacesType = `${Faces}`;

import unpressed from 'assets/face/face_unpressed.svg';
import lose from 'assets/face/face_lose.svg';
import win from 'assets/face/face_win.svg';

const FaceAsserts = {
  [Faces.UNPRESSED]: unpressed,
  [Faces.LOSE]: lose,
  [Faces.WIN]: win,
};

interface FaceProps {
  type: FacesType;
}

export * from './Face';
export { Faces, FaceAsserts };
export type { FacesType, FaceProps };
