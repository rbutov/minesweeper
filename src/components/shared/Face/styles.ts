import styled from 'styled-components';
import { FaceAsserts, FacesType } from '.';

const FaceContainer = styled.div<{ type: FacesType }>`
  width: 39px;
  height: 39px;
  background-size: 100% 100%;

  ${({ type }) => `background-image: url(${FaceAsserts[type]})`};
`;

export { FaceContainer };
