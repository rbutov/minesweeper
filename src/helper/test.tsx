import React from 'react';
import { Store } from 'redux';
import { Provider } from 'react-redux';
import { ReactWrapper, ShallowWrapper } from 'enzyme';
import { act } from 'react-dom/test-utils';
import { confStore } from 'store';

type Wrapper<P = unknown, S = unknown> = ShallowWrapper<P, S> | ReactWrapper<P, S>;
type MountFnType<P, S> = (
  node: React.ReactElement<P>,
  options?: Record<string, unknown>
) => Wrapper<P, S>;

const wrapAppComponent = <P, S>(
  renderer: MountFnType<P, S>,
  element: JSX.Element,
  other?: {
    store?: Store;
  }
): Wrapper<P, S> => {
  const coreStore = confStore();

  return renderer(<Provider store={other?.store ? other.store : coreStore}>{element}</Provider>);
};

const waitForComponentToPaint = async <P, S>(
  wrapper: Wrapper<P, S>,
  other?: { delay?: number; props?: Pick<P, keyof P> }
) => {
  await act(async () => {
    await runAllPromises();
    await new Promise((resolve) => setTimeout(resolve, other?.delay ? other.delay : 0));
    wrapper.setProps({ ...other?.props });
    wrapper.update();
  });
};

const runAllPromises = async () => {
  return new Promise<void>((resolve) => {
    setTimeout(resolve, 0);
  });
};

export { wrapAppComponent, waitForComponentToPaint };
export type { Wrapper };
