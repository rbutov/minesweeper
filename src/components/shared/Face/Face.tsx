import React from 'react';
import { initField } from 'actions/fieldActions';
import { useAppDispatch } from 'hooks/useAppDispatch';
import { FaceContainer } from './styles';
import { FaceProps } from '.';

const Face = ({ type }: FaceProps) => {
  const dispatch = useAppDispatch();
  const handleClick = (ev: React.MouseEvent<HTMLElement>) => {
    ev.preventDefault();
    dispatch(initField());
  };

  return (
    <>
      <FaceContainer onClick={handleClick} type={type} />
    </>
  );
};

export { Face };
