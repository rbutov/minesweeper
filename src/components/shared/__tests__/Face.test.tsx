import React from 'react';
import { wrapAppComponent } from 'helpers/test';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Face, Faces } from '../Face';

describe('<Face />', () => {
  it('_render', () => {
    const wrapper = wrapAppComponent(mount, <Face type={Faces.WIN} />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
