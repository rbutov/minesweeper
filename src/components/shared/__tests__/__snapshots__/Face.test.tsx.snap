// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Face /> _render 1`] = `
<Provider
  store={
    Object {
      "@@observable": [Function],
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Face
    type="win"
  >
    <styled.div
      data-testid="face"
      onClick={[Function]}
      type="win"
    >
      <div
        className="sc-gsDKAQ daujPg"
        data-testid="face"
        onClick={[Function]}
        type="win"
      />
    </styled.div>
  </Face>
</Provider>
`;
